moveit_setup_assistant/src/tools/moveit_config_data.cpp

```diff
@@ -554,6 +554,213 @@ bool MoveItConfigData::outputFakeControllersYAML(const std::string& file_path)
+    emitter << YAML::Value << srdf_->groups_[i].name_ + "_controller";
+    emitter << YAML::Key << "joints";
+    emitter << YAML::Value << YAML::BeginSeq;
+
+    // Iterate through the joints
+    for (int j = 0; j < groups[i].size(); j++)
+    {
+      emitter << groups[i][j];
+    }
+    emitter << YAML::EndSeq;
+    emitter << YAML::EndMap;
+  }
+  emitter << YAML::EndSeq;
+
+  emitter << YAML::Key << "constraints";
+  emitter << YAML::Value << YAML::BeginMap;
+  emitter << YAML::Key << "goal_time";
+  emitter << YAML::Value << "5.0";
+
+  emitter << YAML::Key << "groups";
+  emitter << YAML::Value << YAML::BeginSeq;
+
+  // Loop through groups
+  for (int i = 0; i < srdf_->groups_.size(); ++i)
+  {
+    emitter << YAML::BeginMap;
+    emitter << YAML::Key << "name";
+    emitter << YAML::Value << srdf_->groups_[i].name_ + "_controller";
+    emitter << YAML::Key << "joints";
+    emitter << YAML::Value << YAML::BeginSeq;
+
+    // Iterate through the joints
+    for (int j = 0; j < groups[i].size(); j++)
+    {
+      emitter << YAML::BeginMap;
+      emitter << YAML::Key << groups[i][j] << YAML::Value << YAML::BeginMap;
+      emitter << YAML::Key << "trajectory";
+      emitter << YAML::Value << "0.60";
+      emitter << YAML::Key << "goal";
+      emitter << YAML::Value << "0.15";
+      emitter << YAML::EndMap;
+      emitter << YAML::EndMap;
+    }
+    emitter << YAML::EndSeq;
+    emitter << YAML::EndMap;
+  }
+  emitter << YAML::EndSeq;
+
+  emitter << YAML::EndMap;
+  emitter << YAML::EndMap;
+  emitter << YAML::EndMap;
+
+  emitter << YAML::Newline;
+  emitter << YAML::Comment("Group Position Controllers");
+  emitter << YAML::Comment("Allows to send single ROS msg of Float64MultiArray to all joints");
+
+  emitter << YAML::BeginMap;
+  emitter << YAML::Key << "joint_position_controller";
+  emitter << YAML::Value << YAML::BeginMap;
+  emitter << YAML::Key << "type";
+  emitter << YAML::Value << "position_controllers/JointGroupPositionController";
+
+  emitter << YAML::Key << "groups";
+  emitter << YAML::Value << YAML::BeginSeq;
+
+  // Loop through groups
+  for (int i = 0; i < srdf_->groups_.size(); ++i)
+  {
+    emitter << YAML::BeginMap;
+    emitter << YAML::Key << "name";
+    emitter << YAML::Value << srdf_->groups_[i].name_ + "_controller";
+    emitter << YAML::Key << "joints";
+    emitter << YAML::Value << YAML::BeginSeq;
+
+    // Iterate through the joints
+    for (int j = 0; j < groups[i].size(); j++)
+    {
+      emitter << groups[i][j];
+    }
+    emitter << YAML::EndSeq;
+    emitter << YAML::EndMap;
+  }
+
+  emitter << YAML::EndSeq;
+  emitter << YAML::EndMap;
+  emitter << YAML::EndMap;
+
+  std::ofstream output_stream(file_path.c_str(), std::ios_base::trunc);
+  if (!output_stream.good())
+  {
+    ROS_ERROR_STREAM("Unable to open file for writing " << file_path);
+    return false;
+  }
+  output_stream << emitter.c_str();
+  output_stream.close();
+
+  return true;  // file created successfully
```

Comment:
this function is very long and readability is low. can it be separated out into a few smaller functions for each section of the yaml?

Reply:
Sure, no problem.